Bluetooth: controller: push topic-ble-llcp branch into main

kruithofa opened the pull request:
Pushes all work done in the topic-ble-llcp branch into main branch
This is a refactoring of the LL control procedures; the refactored
control procedures are hidden behind a KConfig option and
per default disabled

Goal of the refactoring:
- close issue #15256
- make it easier to add/update control procedures

Refactoring consists in principal of writing explicit state machines for the control procedures.
To reduce the risk of regression errors unit-tests have been added

Following control procedures are implemented:
- Connection  update procedure
- Channel map update procedure
- Encryption procedure
- Feature exchange procedure
- Version exchange procedure
- ACL termination procedure
- Connection parameters request procedure
- LE Ping procedure
- Data Length Update procedure
- PHY update procedure
- Min. nr. Of channels used procedure
- Constant Tone extension request procedure


This is a joined work by the people listed in the signed-off-by
list (in alphabetical order)

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>
Signed-off-by: Erik Brockhoff <erbr@oticon.com>
Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>
Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>
Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>
Signed-off-by: Tommie Skriver <tosk@demant.com>

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

jhedberg commented:
Is there a reason why commit history hasn't been preserved? Gigantic single commits like this are usually not recommended if they can be avoided.

kruithofa commented:
> Is there a reason why commit history hasn't been preserved? Gigantic single commits like this are usually not recommended if they can be avoided.

It was done to avoid having 270 commits, many of them making updates to the same files

jhedberg reviewed (changes requested):
> It was done to avoid having 270 commits, many of them making updates to the same files

Fair enough, however the commit message then seems way too short for such a big addition. It should preferably contain a high-level architectural summary of the changes, the benefits of the new design, etc.

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -208,6 +208,33 @@
 
 #define PKT_BIS_US(octets, mic, phy) PDU_MAX_US((octets), (mic), (phy))
 
+/* EGON TODO: verify if the following lines are correct */
```
What is EGON? 

Have these been verified by now?

kruithofa replied:
See issue #15256
One of the remaining action items is to go through all TODOs and evaluate what is to be done with these

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -0,0 +1,1109 @@
+/*
+ * Copyright (c) 2020 Demant
```
We can also consider if we want to add `2021` to the copyright lines

kruithofa replied:
Yes, we'll keep it in mind when making further changes

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_chmu.c:
```diff
@@ -0,0 +1,322 @@
+
+static void rp_chmu_st_idle(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+{
+	/* TODO */
```
Still valid TODO? 

Empty TODOs usually aren't very hopeful.

kruithofa replied:
We have a See issue #15256
One of the remaining action items is to go through all TODOs and evaluate what is to be done with these

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -0,0 +1,1118 @@
+}
+
+#if defined(CONFIG_BT_PERIPHERAL)
+static uint8_t pu_check_update_ind(struct ll_conn *conn, struct proc_ctx *ctx)
+{
+	/* Both tx and rx PHY unchanged */
+	if (!((ctx->data.pu.c_to_p_phy | ctx->data.pu.p_to_c_phy) & 0x07)) {
+		/* if no phy changes, quit procedure, and possibly signal host */
+		ctx->data.pu.error = BT_HCI_ERR_SUCCESS;
+		return 1;
+	} else {
+		/* if instant already passed, quit procedure with error */
+		if (is_instant_reached_or_passed(ctx->data.pu.instant, pu_event_counter(conn))) {
+			ctx->data.pu.error = BT_HCI_ERR_INSTANT_PASSED;
+			return 1;
+		}
+	}
+	return 0;
+}
```
I suppose 0 == Success, and 1 == failure;

It feels like an uncommon way of handling error values, due to it being a unsigned value and also that the error value is undefined. 

Possibly refactor to return int and use errno or use a boolean value instead

kruithofa replied:
This code also causes CI failure and is refactored

Thalley commented on tests/bluetooth/bsim_bt/_compile_permutate_kconfigs.sh:
```diff
@@ -0,0 +1,91 @@
+# compile_failures will be equal to the number of failed compilations
+let compile_failures=0
+
+for (( i = 0; i < 2**n; i++ )); do
+    ## don't compile for CENTRAL=n AND PERIPHERAL=n
+    if (( (i & 0x3) != 0x3 )); then
+        perm_compile $n $i ${temp_conf_file}
+    fi
+done
```
This is compiling everything sequential, right? I think we could possibly benefit from using the GNU parallel here.

Thalley commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/tests_scripts/gap.test_list:
```diff
@@ -20,7 +20,7 @@ GAP/CONN/ACEP/BV-01-C
 GAP/CONN/ACEP/BV-04-C
 GAP/CONN/DCON/BV-01-C
-GAP/CONN/ENC
+#GAP/CONN/ENC
```
Is this still intentional?

kruithofa replied:
It is intentional. There is no test GAP/CONN/ENC, so this line will probably be removed in the future

Thalley commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/tests_scripts/hci.test_list:
```diff
@@ -16,16 +16,16 @@ HCI/CIN/BV-03-C
-HCI/DSU/BV-06-C
-HCI/GEV/BV-01-C
-HCI/HFC/BV-04-C
+#HCI/DSU/BV-05-C
+HCI/DSU/BV-06-C # [Reset Command received in Master Role]
+# HCI/GEV/BV-01-C
```
Should these still be commented out?

kruithofa replied:
They seem to fail on the refactored control procedures, and require further investigation

Thalley commented on tests/bluetooth/controller/ctrl_hci/testcase.yaml:
```diff
@@ -0,0 +1,6 @@
+    tags: test_framework bluetooth bt_ctrl_hci bt_ull_llcp
+tests:
+  bluetooth.controller.ctrl_hci.test:
+     skip: true
```
Why is this being skipped?

thoh-ot replied:
It's being skipped because It doesn't build anymore, the preparation for up-streaming changed how the interfaces were exposed and we didn't have time to fix this prior to up-streaming.

Thalley commented on tests/bluetooth/controller/ctrl_min_used_chans/src/main.c:
```diff
@@ -0,0 +1,179 @@
+#include <ztest.h>
+#include "kconfig.h"
+
+/* Kconfig Cheats */
```
This comment seems out of place

Thalley commented on tests/bluetooth/controller/ctrl_terminate/src/main.c:
```diff
@@ -0,0 +1,149 @@
+#include <ztest.h>
+#include "kconfig.h"
+
+/* Kconfig Cheats */
```
Ditto

Thalley commented on tests/bluetooth/controller/mock_ctrl/src/ull_periph.c:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Is `periph` the abbreviation we want to go with?

I suggest to write it out complete as `peripheral`.

kruithofa replied:
We'll consider this after this PR is merged into main

kruithofa replied:
Fixed in PR #40458

kruithofa commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -116,9 +116,35 @@ config BT_CTLR_ADVANCED_FEATURES
+	  Use the Bluetooth Low Energy Software Link Layer Legacy Control Procedure implementation.
+
+config BT_LL_SW_SPLIT_LLCP_CHOICE_REFACTORED
+	bool "Refactored [EXPERIMENTAL]"
```
Correct. It could not be done in the topic branch (@sjanc has the details), and will be fixed when this PR is merged in main

kruithofa replied:
'select EXPERIMENTAL' has been added

kruithofa commented on include/bluetooth/hci_vs.h:
```diff
@@ -201,6 +201,15 @@ struct bt_hci_cp_vs_set_usb_transport_mode {
+	uint8_t  min_used_chans;
+} __packed;
+
+
```
Should be fixed now

kruithofa commented on scripts/ci/run_ci.sh:
```diff
@@ -142,6 +159,7 @@ function build_test_file() {
 		test_file_boards_in.txt >> test_file.txt
 }
 
+
```
Should be fixed now

thoh-ot commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/hci_test_app/prj_tst.conf:
```diff
@@ -3,7 +3,9 @@
 CONFIG_BT=y
 CONFIG_BT_HCI_RAW=y
 CONFIG_BT_PERIPHERAL=y
+CONFIG_BT_BROADCASTER=y
```
Why did we add this?

kruithofa replied:
Because it was added in the main branch

thoh-ot commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/hci_test_app/prj_tst.conf:
```diff
@@ -3,7 +3,9 @@
 CONFIG_BT_PERIPHERAL=y
+CONFIG_BT_BROADCASTER=y
 CONFIG_BT_CENTRAL=y
+CONFIG_BT_OBSERVER=y
```
Why did we add this?

kruithofa replied:
In this file it needs to be removed; I'll do that

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_peripheral.c:
```diff
@@ -172,6 +180,11 @@ void ull_periph_setup(struct node_rx_hdr *rx, struct node_rx_ftr *ftr,
 		win_delay_us = WIN_DELAY_LEGACY;
 	}
 
+#if (!defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY))
```
nitpick: superfluous ()

kruithofa replied:
The conditional compile issues will be fixed after this PR is merged in main

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_peripheral.c:
```diff
@@ -30,10 +30,14 @@
+
+#if !defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY)
+#include "ull_tx_queue.h"
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_peripheral.c:
```diff
@@ -46,6 +50,10 @@
 
+#if (!defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY))
+#include "ll_sw/ull_llcp.h"
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_peripheral.c:
```diff
@@ -313,8 +326,13 @@ void ull_periph_setup(struct node_rx_hdr *rx, struct node_rx_ftr *ftr,
+#if defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY)
 	max_tx_time = lll->max_tx_time;
 	max_rx_time = lll->max_rx_time;
+#else
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_peripheral.c:
```diff
@@ -313,8 +326,13 @@ void ull_periph_setup(struct node_rx_hdr *rx, struct node_rx_ftr *ftr,
+#else
+	max_tx_time = lll->dle.local.max_tx_time;
+	max_rx_time = lll->dle.local.max_rx_time;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -26,6 +26,10 @@
 
+#if !defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY)
+#include "ull_tx_queue.h"
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -38,10 +42,12 @@
 			    uint32_t *ticks_to_offset_next,
 			    uint16_t conn_interval, uint8_t *offset_max,
 			    uint8_t *win_offset);
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -194,7 +200,13 @@ void ull_sched_mfy_win_offset_use(void *param)
+	 */
+#if defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY)
 	uint16_t win_offset;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -214,18 +230,30 @@ void ull_sched_mfy_win_offset_use(void *param)
 
 	/* move to offset calculated state */
 	conn->llcp_cu.state = LLCP_CUI_STATE_OFFS_RDY;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -214,18 +230,30 @@ void ull_sched_mfy_win_offset_use(void *param)
+#if defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY)
 	uint8_t offset_max = 6U;
+	struct ll_conn *conn = param;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -242,19 +270,28 @@ void ull_sched_mfy_free_win_offset_calc(void *param)
-	uint32_t ticks_to_offset;
 	uint16_t win_offset_s;
+	uint32_t ticks_to_offset;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on tests/bluetooth/controller/common/include/helper_features.h:
```diff
@@ -0,0 +1,249 @@
+ */
+#if defined(CONFIG_BT_CTLR_LE_ENC)
+#define FEAT_ENCODED 0x01
+#else
```
missing /* ??? */ throughout

erbr-ot commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -480,7 +480,11 @@ void lll_conn_rx_pkt_set(struct lll_conn *lll)
 #if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+#ifdef CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY
 	max_rx_octets = lll->max_rx_octets;
+#else
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -480,7 +480,11 @@ void lll_conn_rx_pkt_set(struct lll_conn *lll)
 	max_rx_octets = lll->max_rx_octets;
+#else
+	max_rx_octets = lll->dle.eff.max_rx_octets;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -524,7 +528,11 @@ void lll_conn_tx_pkt_set(struct lll_conn *lll, struct pdu_data *pdu_data_tx)
 #if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+#ifdef CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY
 	max_tx_octets = lll->max_tx_octets;
+#else
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -524,7 +528,11 @@ void lll_conn_tx_pkt_set(struct lll_conn *lll, struct pdu_data *pdu_data_tx)
 	max_tx_octets = lll->max_tx_octets;
+#else
+	max_tx_octets = lll->dle.eff.max_tx_octets;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll_conn.c:
```diff
@@ -427,7 +429,11 @@ void lll_conn_rx_pkt_set(struct lll_conn *lll)
 #if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+#ifdef CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY
 	max_rx_octets = lll->max_rx_octets;
+#else
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll_conn.c:
```diff
@@ -427,7 +429,11 @@ void lll_conn_rx_pkt_set(struct lll_conn *lll)
 	max_rx_octets = lll->max_rx_octets;
+#else
+	max_rx_octets = lll->dle.eff.max_rx_octets;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll_conn.c:
```diff
@@ -461,7 +467,11 @@ void lll_conn_tx_pkt_set(struct lll_conn *lll, struct pdu_data *pdu_data_tx)
 #if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+#ifdef CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY
 	max_tx_octets = lll->max_tx_octets;
+#else
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll_conn.c:
```diff
@@ -461,7 +467,11 @@ void lll_conn_tx_pkt_set(struct lll_conn *lll, struct pdu_data *pdu_data_tx)
 	max_tx_octets = lll->max_tx_octets;
+#else
+	max_tx_octets = lll->dle.eff.max_tx_octets;
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -45,6 +45,9 @@
 #include "ull_sync_types.h"
+#if !defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY)
+#include "ull_tx_queue.h"
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -957,7 +966,17 @@ uint8_t ll_adv_enable(uint8_t enable)
 							  lll->phy_s));
 #endif /* CONFIG_BT_CTLR_ADV_EXT */
 #endif /* CONFIG_BT_CTLR_PHY */
+#endif
```
missing /* CONFIG_BT_CTLR_DATA_LENGTH */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -957,7 +966,17 @@ uint8_t ll_adv_enable(uint8_t enable)
+#if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+#if defined(CONFIG_BT_CTLR_PHY) && defined(CONFIG_BT_CTLR_ADV_EXT)
+		const uint8_t phy = lll->phy_s;
+#else
```
missing /* CONFIG_BT_CTLR_PHY && CONFIG_BT_CTLR_ADV_EXT */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -957,7 +966,17 @@ uint8_t ll_adv_enable(uint8_t enable)
+		const uint8_t phy = lll->phy_s;
+#else
+		const uint8_t phy = PHY_1M;
+#endif
```
missing /* CONFIG_BT_CTLR_PHY && CONFIG_BT_CTLR_ADV_EXT */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_chan_internal.h:
```diff
@@ -6,3 +6,6 @@
 uint8_t ull_chan_map_get(uint8_t *const chan_map);
+#if !defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY)
+void ull_chan_map_set(uint8_t const *const chan_map);
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -509,11 +582,20 @@ uint8_t ll_feature_req_send(uint16_t handle)
+	if (err) {
+		return err;
+	}
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -533,11 +616,20 @@ uint8_t ll_version_ind_send(uint16_t handle)
+	if (err) {
+		return err;
+	}
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1875,6 +2130,27 @@ static struct node_tx *tx_ull_dequeue(struct ll_conn *conn, struct node_tx *tx)
 
 	return tx;
 }
+#else
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -7120,3 +7421,416 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate)
+	     conn->lll.dle.local.max_tx_time != max_time_min)) {
+		conn->lll.dle.update = 1;
+	} else
+#endif
```
missing /* CONFIG_BT_CTLR_PHY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -7120,3 +7421,416 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate)
+#if defined(CONFIG_BT_CTLR_PHY)
+	const uint16_t max_time_min = PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, phy);
+	const uint16_t max_time_max = PDU_DC_MAX_US(LL_LENGTH_OCTETS_RX_MAX, phy);
+#endif
```
missing /* CONFIG_BT_CTLR_PHY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -7120,3 +7421,416 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate)
+		conn->lll.dle.eff.max_rx_time = eff_rx_time;
+		dle_changed = 1;
+	}
+#else
```
missing /* CONFIG_BT_CTLR_PHY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -7120,3 +7421,416 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate)
+#else
+	conn->lll.dle.eff.max_rx_time = PDU_DC_MAX_US(eff_rx_octets, PHY_1M);
+	conn->lll.dle.eff.max_tx_time = PDU_DC_MAX_US(eff_tx_octets, PHY_1M);
+#endif
```
missing /* CONFIG_BT_CTLR_PHY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -7120,3 +7421,416 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate)
+		phy_select = PHY_CODED;
+		/* If not, max times should be defined by 1M timing */
+	}
+#endif
```
missing /* CONFIG_BT_CTLR_PHY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn_types.h:
```diff
@@ -345,6 +346,194 @@ struct ll_conn {
 	struct lll_df_conn_rx_params df_rx_params;
 #endif /* CONFIG_BT_CTLR_DF_CONN_CTE_REQ */
 };
+#else
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_filter.c:
```diff
@@ -30,6 +30,10 @@
 
+#if (!defined(CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY))
+#include "ll_sw/ull_tx_queue.h"
+#endif
```
missing /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */
and superfluous () on define

erbr-ot reviewed (commented):
Sorry about all the nitpicks.
Otherwise I think things look fine

Commit:
Bluetooth: controller: push topic branch to main

Pushes all work done in the topic-ble-llcp branch into main branch
This is a refactoring of the LL control procedures; the refactored
control procedures are hidden behind a KConfig option and
per default disabled

Goal of the refactoring:

close issue Link Layer Control Procedure overhaul #15256
make it easier to add/update control procedures
Refactoring consists in principal of writing explicit state machines
for the control procedures.
To reduce the risk of regression errors unit-tests have been added

Following control procedures are implemented:

Connection update procedure
Channel map update procedure
Encryption procedure
Feature exchange procedure
Version exchange procedure
ACL termination procedure
Connection parameters request procedure
LE Ping procedure
Data Length Update procedure
PHY update procedure
Min. nr. Of channels used procedure
Constant Tone extension request procedure

This is a joined work by the people listed in the signed-off-by
list (in alphabetical order)

Signed-off-by: Andries Kruithof Andries.Kruithof@nordicsemi.no
Signed-off-by: Erik Brockhoff erbr@oticon.com
Signed-off-by: Piotr Pryga piotr.pryga@nordicsemi.no
Signed-off-by: Szymon Janc szymon.janc@codecoup.pl
Signed-off-by: Thomas Ebert Hansen thoh@oticon.com
Signed-off-by: Tommie Skriver tosk@demant.com

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

jhedberg commented:
@kruithofa please include the stuff you added to the PR description also into the commit message since that's all that will be visible to people looking at the commit history (there's no easy way to track down PRs based on a commit).

erbr-ot reviewed (approved):
Here goes nothing!

thoh-ot commented:
What's up with the CI?

kruithofa commented:
> What's up with the CI?

I'd like to know as well; failures seem to be in areas which are not touched by this PR

sjanc commented:
Maybe just unlucky base commit?

stephanosio reviewed (dismissed):
```
INFO    - 1/1 qemu_cortex_a53_smp       tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace FAILED Timeout (qemu)
```

`tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace` finishes in approx. 2.5s on the latest main, but it times out on this PR branch.

You can reproduce this locally using the following:
```
scripts/twister -v -N -p qemu_cortex_a53_smp -T tests/kernel/mem_protect/userspace
```

This PR must be making some change that is causing this.

carlescufi commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -373,10 +376,6 @@ static int run_test(struct unit_test *test)
 		run_test_functions(test);
 	}
 
-	phase = TEST_PHASE_TEARDOWN;
```
This change is causing a CI build failure. In particular, to reproduce:

```
$ scripts/twister -v -p qemu_cortex_a53_smp -T tests/kernel/mem_protect/userspace/
```

kruithofa replied:
working on it

carlescufi reviewed (approved):
LGTM now

cvinayak commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -276,6 +279,8 @@ struct node_rx_ftr {
 				*/
 		void *aux_ptr;
 		uint8_t aux_phy;
+		uint8_t aux_sched;
```
Is this related to LLCP work?

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -47,6 +47,7 @@
 #define LOG_MODULE_NAME bt_ctlr_lll_scan_aux
 #include "common/log.h"
 #include <soc.h>
+#include <ull_scan_types.h>
```
Why is ULL types referenced in LLL ?

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -693,7 +700,6 @@ static void isr_rx(struct lll_scan *lll, struct lll_scan_aux *lll_aux,
 		irkmatch_ok = radio_ar_has_match();
 		irkmatch_id = radio_ar_match_get();
 		rssi_ready = radio_rssi_is_ready();
-		phy_aux_flags_rx = radio_phy_flags_rx_get();
```
Why is this removed?

kruithofa replied:
By accident. Reintroduced in PR #40510

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -26,6 +26,7 @@
 #include "lll_vendor.h"
 #include "lll_clock.h"
 #include "lll_chan.h"
+#include "lll/lll_df_types.h"
```
Why add this?  If required, then use `#include "lll_df_types.h"`

kruithofa replied:
fixed

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -453,6 +454,12 @@ static int prepare_cb(struct lll_prepare_param *p)
 	}
 #endif /* CONFIG_BT_CENTRAL */
 
+	/* Initialize scanning state */
+	lll->state = 0U;
+
+	/* Reset Tx/rx count */
+	trx_cnt = 0U;
+
```
Why is this added?

kruithofa replied:
as part of LLCP refactoring

cvinayak commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll_adv.c:
```diff
@@ -31,6 +31,7 @@
 #include "lll_adv_types.h"
 #include "lll_adv.h"
 #include "lll_adv_pdu.h"
+#include "lll_df_types.h"
```
Why add this?

cvinayak commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -208,6 +208,33 @@
 
 #define PKT_BIS_US(octets, mic, phy) PDU_MAX_US((octets), (mic), (phy))
 
+/* TODO: verify if the following lines are correct */
+/* Extra bytes for enqueued node_rx metadata: rssi (always), resolving
+ * index, directed adv report, and mesh channel and instant.
+ */
+#define PDU_AC_SIZE_RSSI 1
+#if defined(CONFIG_BT_CTLR_PRIVACY)
+#define PDU_AC_SIZE_PRIV 1
+#else
+#define PDU_AC_SIZE_PRIV 0
+#endif /* CONFIG_BT_CTLR_PRIVACY */
+#if defined(CONFIG_BT_CTLR_EXT_SCAN_FP)
+#define PDU_AC_SIZE_SCFP 1
+#else
+#define PDU_AC_SIZE_SCFP 0
+#endif /* CONFIG_BT_CTLR_EXT_SCAN_FP */
+#if defined(CONFIG_BT_HCI_MESH_EXT)
+#define PDU_AC_SIZE_MESH 5
+#else
+#define PDU_AC_SIZE_MESH 0
+#endif /* CONFIG_BT_HCI_MESH_EXT */
+
+#define PDU_AC_LL_SIZE_EXTRA (PDU_AC_SIZE_RSSI + \
+			      PDU_AC_SIZE_PRIV + \
+			      PDU_AC_SIZE_SCFP + \
+			      PDU_AC_SIZE_MESH)
+
+
```
Why add this?

kruithofa replied:
It was originally needed, but I see the use of PDU_AC_LL_SIZE_EXTRA was removed in the main branch a while ago (but obviously not in the topic branch). I'll create a PR to remove this

kruithofa replied:
see PR #40503 for removing these

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_types.h:
```diff
@@ -116,3 +116,21 @@ struct lll_df_conn_rx_params {
+/* @brief Structure to store data required to prepare LE Connection IQ Report event or LE
+ * Connectionless IQ Report event.
+ *
+ * TODO (ppryga): use struct cte_conn_iq_report in connected mode. Members are exactly the same as
```
`ppryga`  ?

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -34,11 +34,11 @@
 
 #include "ull_internal.h"
 #include "ull_adv_internal.h"
+#include "ull_scan_types.h"
```
Any reason to move the include to group of internal includes?

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -29,6 +29,7 @@
 #include "lll_chan.h"
 #include "lll_scan.h"
 #include "lll/lll_df_types.h"
+#include "lll_conn.h"
```
why lll_conn.h is include for periodic advertising sync implementation?

cvinayak commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/gatt_test_app/prj.conf:
```diff
@@ -26,3 +26,5 @@ CONFIG_BT_CTLR_RX_BUFFERS=3
 
 # To make DEVICE Name writable...
 CONFIG_BT_DEVICE_NAME_DYNAMIC=y
+
+CONFIG_BT_LL_SW_LLCP_LEGACY=y
```
Why no GATT tests on new LLCP? i.e. no prj_dut_llcp.conf?

kruithofa replied:
The question is if GATT (and GAP) tests are effected by LLCP. For sake of completeness I can add a test config for the new LLCP

kruithofa replied:
see PR #40506 for running GATT and GAP tests with new LLCP

cvinayak commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/tests_scripts/hci.llcp.test_list:
```diff
@@ -0,0 +1,31 @@
+# Copyright 2019 Oticon A/S
+# SPDX-License-Identifier: Apache-2.0
+
+HCI/CCO/BV-07-C
+HCI/CCO/BV-09-C # [Handling LE Set Data Length Command]
+HCI/CCO/BV-10-C
+HCI/CCO/BV-11-C
+HCI/CCO/BV-12-C
+HCI/CCO/BV-13-C
+HCI/CCO/BV-14-C
+HCI/CCO/BV-15-C
+HCI/CCO/BV-18-C
+HCI/CFC/BV-02-C
+HCI/CIN/BV-01-C
+HCI/CIN/BV-03-C
+HCI/CIN/BV-04-C
+HCI/CIN/BV-06-C
+HCI/CIN/BV-09-C
+HCI/CM/BV-01-C # [Handling LE Read Peer Resolvable Address Command]
+HCI/CM/BV-02-C # [Handling LE Read Local Resolvable Address Command]
+HCI/CM/BV-03-C # [Handling LE Read PHY Command]
+HCI/DDI/BI-02-C
+HCI/DDI/BV-03-C
+HCI/DDI/BV-04-C
+HCI/DSU/BV-02-C
+HCI/DSU/BV-03-C # [Reset Command received in Slave Role]
+HCI/DSU/BV-04-C
+#HCI/DSU/BV-05-C
+HCI/DSU/BV-06-C # [Reset Command received in Master Role]
+# HCI/GEV/BV-01-C
+HCI/HFC/BV-04-C # [Events enabled by LE Set Event Mask Command]
```
Are these additional tests?

kruithofa replied:
They are the same tests as for legacy, but for the new llcp

cvinayak commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/tests_scripts/hci.test_list:
```diff
@@ -27,5 +27,5 @@ HCI/DSU/BV-03-C
 HCI/DSU/BV-05-C
 HCI/DSU/BV-06-C
-HCI/GEV/BV-01-C
+#HCI/GEV/BV-01-C
```
Why is this disabled?

kruithofa replied:
We saw failures on this in the topic branch, but is on the todo list and must be fixed

cvinayak commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/tests_scripts/ll.set1.llcp.test_list:
```diff
@@ -0,0 +1,64 @@
+#LL/CON/MAS/BI-06-C #currently failing, to be investigated
+LL/CON/MAS/BV-03-C
+LL/CON/MAS/BV-04-C
+LL/CON/MAS/BV-05-C
+LL/CON/MAS/BV-07-C
+LL/CON/MAS/BV-08-C
+LL/CON/MAS/BV-09-C
+LL/CON/MAS/BV-13-C
+LL/CON/MAS/BV-20-C
+LL/CON/MAS/BV-21-C
+LL/CON/MAS/BV-23-C
+LL/CON/MAS/BV-24-C
+LL/CON/MAS/BV-25-C
+LL/CON/MAS/BV-26-C
+#LL/CON/MAS/BV-27-C #currently failing, to be investigated
+LL/CON/MAS/BV-29-C
+LL/CON/MAS/BV-30-C
+LL/CON/MAS/BV-34-C
+LL/CON/MAS/BV-35-C
+LL/CON/MAS/BV-41-C
+LL/CON/MAS/BV-43-C
+#LL/CON/MAS/BV-73-C # requires update of EDTT due to change in DLE algorithm
+#LL/CON/MAS/BV-74-C # requires update of EDTT due to change in DLE algorithm
+#LL/CON/MAS/BV-76-C # requires update of EDTT due to change in DLE algorithm
+#LL/CON/MAS/BV-77-C # requires update of EDTT due to change in DLE algorithm
+#LL/CON/SLA/BI-08-C # requires update of EDTT due to change in DLE algorithm
+LL/CON/SLA/BV-04-C
+LL/CON/SLA/BV-05-C
+LL/CON/SLA/BV-06-C
+LL/CON/SLA/BV-10-C
+LL/CON/SLA/BV-11-C
+LL/CON/SLA/BV-12-C
+LL/CON/SLA/BV-14-C
+LL/CON/SLA/BV-19-C
+LL/CON/SLA/BV-20-C
+LL/CON/SLA/BV-22-C
+LL/CON/SLA/BV-24-C
+LL/CON/SLA/BV-25-C
+LL/CON/SLA/BV-26-C
+LL/CON/SLA/BV-27-C
```
Are these additional tests compare to being tested for legacy LLCP?

kruithofa replied:
Same tests as for legacy, but for new LLCP

cvinayak commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/tests_scripts/ll.set2.llcp.test_list:
```diff
@@ -0,0 +1,62 @@
+LL/DDI/ADV/BV-15-C
+LL/DDI/ADV/BV-16-C
+LL/DDI/ADV/BV-17-C
+LL/DDI/ADV/BV-18-C
+LL/DDI/ADV/BV-19-C
+LL/DDI/ADV/BV-20-C
+LL/DDI/SCN/BV-01-C
+LL/DDI/SCN/BV-02-C
+LL/DDI/SCN/BV-03-C
+LL/DDI/SCN/BV-04-C
+LL/DDI/SCN/BV-05-C
+LL/DDI/SCN/BV-10-C
+LL/DDI/SCN/BV-11-C
+LL/DDI/SCN/BV-12-C
+LL/DDI/SCN/BV-13-C
+LL/DDI/SCN/BV-14-C
+LL/DDI/SCN/BV-15-C
+LL/DDI/SCN/BV-16-C
+LL/DDI/SCN/BV-17-C
+LL/DDI/SCN/BV-18-C
+LL/DDI/SCN/BV-26-C
+LL/DDI/SCN/BV-28-C
+LL/SEC/ADV/BV-02-C
+LL/SEC/ADV/BV-03-C
+LL/SEC/ADV/BV-04-C
+LL/SEC/ADV/BV-05-C
+LL/SEC/ADV/BV-06-C
+LL/SEC/ADV/BV-08-C
+LL/SEC/ADV/BV-09-C
+LL/SEC/ADV/BV-10-C
+LL/SEC/ADV/BV-11-C
+LL/SEC/ADV/BV-12-C
+LL/SEC/ADV/BV-13-C
+LL/SEC/ADV/BV-14-C
+LL/SEC/ADV/BV-15-C
+LL/SEC/ADV/BV-16-C
+LL/SEC/ADV/BV-17-C
+LL/SEC/ADV/BV-18-C
+LL/SEC/ADV/BV-20-C
+LL/SEC/SCN/BV-01-C
```
Are these additional tests compare to being tested for legacy LLCP?

kruithofa replied:
same tests as for legacy but for new LLCP

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_internal.h:
```diff
@@ -31,6 +31,7 @@ static inline void ull_hdr_init(struct ull_hdr *hdr)
 	hdr->disabled_cb = hdr->disabled_param = NULL;
 }
 
+void ll_tx_ack_put(uint16_t handle, struct node_tx *node_tx);
```
Remove this, as it is present in ull_conn_internal.h

cvinayak replied:
Created https://github.com/zephyrproject-rtos/zephyr/pull/40487

tejlmand commented:
Seems I was never invited to the party.

The initial looks at things like:
https://github.com/kruithofa/zephyr/blob/8a39814cbf5f450e2e317890f079e875f7b0abf5/tests/bluetooth/controller/common/defaults_cmake.txt
https://github.com/kruithofa/zephyr/blob/8a39814cbf5f450e2e317890f079e875f7b0abf5/tests/bluetooth/controller/common/defaults_cmake.txt#L59

seems to be contrary to everything else we do in build system and test build code in general.

Not sure I understand why it was done this way, and as all history has been squashed, it's extremely hard to know the reasons for those decisions.
@kruithofa @erbr-ot FYI

ppryga commented:
> seems to be contrary to everything else we do in build system and test build code in general.

@tejlmand could you elaborate on this?

tejlmand commented:
> @tejlmand could you elaborate on this?

@ppryga sure.

First, we use Kconfig to configure the build system. This creates `autoconf.h` that are added as compile option (`-imacro`) to all Zephyr library sources.

Now, with this code:
https://github.com/zephyrproject-rtos/zephyr/blob/8a39814cbf5f450e2e317890f079e875f7b0abf5/tests/bluetooth/controller/common/defaults_cmake.txt#L58-L61
you completely bypass Kconfig and autoconf.h by providing a custom `kconfig.h` file.
https://github.com/zephyrproject-rtos/zephyr/blob/8a39814cbf5f450e2e317890f079e875f7b0abf5/tests/bluetooth/controller/mock_ctrl/include/kconfig.h
which you then suddenly sometimes decides to overwrite with an extra header file:
https://github.com/zephyrproject-rtos/zephyr/blob/8a39814cbf5f450e2e317890f079e875f7b0abf5/tests/bluetooth/controller/ctrl_data_length_update/testcase.yaml#L12
which undefines previously set defines:
https://github.com/zephyrproject-rtos/zephyr/blob/8a39814cbf5f450e2e317890f079e875f7b0abf5/tests/bluetooth/controller/ctrl_data_length_update/src/kconfig_override.h#L12-L14

This is really not a pattern that should spread, and it should be cleaned up.
You use `add_definitions()` for setting compile options. A function superseded by newer alternatives:
https://cmake.org/cmake/help/latest/command/add_definitions.html
> **Note**: This command has been superseded by alternatives: 

https://github.com/zephyrproject-rtos/zephyr/blob/8a39814cbf5f450e2e317890f079e875f7b0abf5/tests/bluetooth/controller/common/defaults_cmake.txt#L60

and I assume you haven't checked whether this code is vulnerable to CMake deduplication of flags ?
Hidden surprises can be waiting in future, that can make things hard to debug.

Secondly, including sources using glob pattern is discouraged by CMake.
And as you are anyway specifying all files, then why use a glob pattern ?
https://github.com/zephyrproject-rtos/zephyr/blob/8a39814cbf5f450e2e317890f079e875f7b0abf5/tests/bluetooth/controller/common/defaults_cmake.txt#L19

This is just for a start.
Haven't reviewed the code in details.

ppryga commented:
@tejlmand thanks. 
I didn't contribute to this PR much and didn't "invent" the approach but I think that you are welcome to provide a way to build unit tests for a subset of source files in Bluetooth Controller. 
I do not know if there is a well established approach to unit testing (in an isolation) of subsets of Zephyr modules. 
This was an approach that was taken for start and it serves its purpose. 
Now it could be updated with better solution.

thoh-ot commented:
@tejlmand My comments as one of the original inventors of this...

We didn't bypass the Kconfig system as it was never there to begin with, because when using the `find_package(ZephyrUnittest HINTS $ENV{ZEPHYR_BASE})` you don't get the `cmake/app/boilerplate.cmake` as boilerplate instead you get `subsys/testsuite/unittest.cmake` and that file doesn't have Kconfig -> `autoconf.h` support.

As some of the Bluetooth controller code is dependent on various Kconfig settings, doing unit tests of parts of it was not so easy, so instead we choose to fake it, giving us 100% control over what the "faked" `autoconf.h` would contain.

Regarding the `add_definitions()` I wasn't aware - and it can be found in other places also. But that's a simple change to fix.

Regarding the 'FILE(GLOB' i agree we got something wrong, but then again there are 530 other files also using glob pattern mechanism.

tejlmand commented:
> We didn't bypass the Kconfig system as it was never there to begin with, because when using the find_package(ZephyrUnittest HINTS $ENV{ZEPHYR_BASE}) you don't get the cmake/app/boilerplate.cmake as boilerplate instead you get subsys/testsuite/unittest.cmake and that file doesn't have Kconfig -> autoconf.h support.

Fair enough, I didn't notice that part.
Probably I would have if I had done a proper review, but I just happened to notice this as part of the #41886 review.

In this case, I believe we need to see if we can improve on the ZephyrUnittest package.
I believe #41301 will allow us to improve in this area as it suddenly allows us to just load parts of build system needed for unittesting. In this case Kconfig could be a good candidate.

> but then again there are 530 other files also using glob pattern mechanism.

yes, but for populating only the `app` lib in toplevel CMakeLists.txt files for samples and tests.
Not used for subsystem / driver libraries, and the like.

This falls somewhat in between, as it is populating same lib, but not done directly in toplevel CMakeLists file.
Those sources are more akin to sources added to subsystems / driver libs.

Btw, `GLOB` is discouraged by CMake for source files:
https://cmake.org/cmake/help/latest/command/file.html#filesystem
> Note: We do not recommend using GLOB to collect a list of source files from your source tree. ...
